design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,777 @@
+# End-to-end Testing in Crossplane
+
+* Owner: Predrag Knežević (@pedjak), Lovro Sviben (@lsviben)
+* Reviewers: @negz, TBD
+* Status: Draft
+
+## Background
+
+As a step in maturing Crossplane, we should be thinking about setting up
+processes and methodology that would enable automated end-to-end testing,
+covering all user-facing functionalities.
+
+Currently, as reported in [this issue], there seems to be a lack of e2e tests
+for Crossplane, impacting directly confidence in changes submitted in PRs and
+the quality of published releases.
+
+The following repos contains some sort of e2e tests:
+
+* [crossplane/test], last test change was 9 month ago
+* Crossplane repo itself [crossplane/crossplane e2e], last update was 10 months
+  ago
+* [crossplane/conformance], last change in June, 2021
+
+Apart for the stale development, the existing tests:
+
+* Cover only tiny fraction of all user-facing features
+* Do not cover any error cases
+* Are reported to be flaky sometimes
+
+The tests are written in pure Go, requesting to keep the number of library
+dependencies at minimum.
+
+Contrary to unit tests, end-to-end testing requires very often complex test 
+fixtures (e.g. installing Crossplane, providers, reaching some state in the
+cluster) before we can focus on validating a particular system behavior.
+Defining the fixture only with the help of Go standard library can be
+time-consuming, requiring to write a significant amount of boiler-plate code,
+hence reducing test readability and maintenance, and eventually not very
+motivating for creating new tests. In order to improve the readability, one
+would start looking for patterns and either eventually end up with a custom
+framework or adopt an existing one.
+
+**_NOTE_**: The document uses end-to-end testing term interchangeably
+with [system testing].
```

Comment:
I know we use the term "E2E" in https://github.com/crossplane/crossplane/issues/4013, but I don't think we need to be too strict about this. I get the sense that in most cloud native projects "E2E testing" often includes integration testing and isn't strictly (whole) system testing.

Reply:
What would be reason for relaxing the term? In Crossplane, integration testing is already covered in [this proposal](https://github.com/crossplane/crossplane/blob/master/design/design-doc-integration-tests.md).

Reply:
That proposal never went anywhere. We don't have any meaningful integration tests.

What we need is something higher level than unit tests that tests code as more of a black box than unit tests do. I think we can be pragmatic as to whether those new tests are integration tests, system tests, or a mix of both.

Reply:
> I think we can be pragmatic as to whether those new tests are integration tests, system tests, or a mix of both.

Pragmatic can mean many things, but it is orthogonal to  being ambiguous. IMHO there is no one-size-fits-all solution for testing - each sort of tests requires specialized abstractions, because they are executed on different levels and expectations.

For me pragmatic means choosing a right tool for a task. As we can see, Crossplane unit tests are alive and kicking, but [the integration testing proposal](https://github.com/crossplane/crossplane/blob/master/design/design-doc-integration-tests.md) did not get traction. That a strong signal that we would need to approach the problem differently.

Reply:
> That is a strong signal that we would need to approach the problem differently.

I disagree. The problem we've had in the past was not choice of framework, but rather bandwidth to prioritize actually writing the tests. Picking any framework is pointless if we can't form and execute a plan to bootstrap the test coverage. If every new PR to the project sees there are no tests, no one is motivated to add tests for their new code.